pkg/server/cstorvolumeconfig/backup_endpoint.go

```diff
@@ -0,0 +1,545 @@
+/*
+Copyright 2020 The OpenEBS Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cstorvolumeconfig
+
+import (
+	"net/http"
+
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	cstorapis "github.com/openebs/api/pkg/apis/cstor/v1"
+	openebsapis "github.com/openebs/api/pkg/apis/openebs.io/v1alpha1"
+	cstortypes "github.com/openebs/api/pkg/apis/types"
+	clientset "github.com/openebs/api/pkg/client/clientset/versioned"
+	"github.com/openebs/api/pkg/util"
+	snapshot "github.com/openebs/cstor-operators/pkg/snapshot"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	k8serror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+)
+
+// backupAPIOps holds the clients, http request and
+// response
+type backupAPIOps struct {
+	req          *http.Request
+	resp         http.ResponseWriter
+	k8sclientset kubernetes.Interface
+	clientset    clientset.Interface
+	snapshotter  snapshot.Snapshotter
+}
+
+var (
+	// openebsNamespace is the namespace where openebs is deployed
+	openebsNamespace string
+)
+
+// backupV1alpha1SpecificRequest deals with backup API requests
+func (s *HTTPServer) backupV1alpha1SpecificRequest(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	backupOp := &backupAPIOps{
+		req:          req,
+		resp:         resp,
+		k8sclientset: s.cvcServer.kubeclientset,
+		clientset:    s.cvcServer.clientset,
+		snapshotter:  s.cvcServer.snapshotter,
+	}
+
+	switch req.Method {
+	case "POST":
+		return backupOp.create()
+	case "GET":
+		klog.Infof("Got backup GET request")
+		return backupOp.get()
+	case "DELETE":
+		klog.Infof("Got backup DELETE request")
+		return backupOp.delete()
+	}
+	return nil, CodedError(405, ErrInvalidMethod)
+}
+
+// Create is http handler which handles backup create request
+func (bOps *backupAPIOps) create() (interface{}, error) {
+	backUp := &openebsapis.CStorBackup{}
+
+	err := decodeBody(bOps.req, backUp)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := backupCreateRequestValidations(backUp); err != nil {
+		return nil, err
+	}
+	klog.Infof("Requested to create backup for volume %s/%s remoteBackup: %t", backUp.Namespace, backUp.Spec.VolumeName, !backUp.Spec.LocalSnap)
+
+	// TODO: Move this to interface so that we can mock
+	// snapshot calls
+	snapshot := snapshot.Snapshot{
+		VolumeName:   backUp.Spec.VolumeName,
+		SnapshotName: backUp.Spec.SnapName,
+		Namespace:    getOpenEBSNamespace(),
```

Comment:
This seems repetitive call. Moving it to backupAPIOps will be better, since openebs namespace is fetched from env variable. This will reduce the serve time for backup request.

Reply:
basically I am storing that value in memory once after reading it. So it will not be time taking a call. It will variable has some value then it will return that value.

Reply:
If we can avoid this function call then better to avoid it. This is mainly for optimization.

Reply:
Done